src/hardware/voodoo.cpp

```diff
@@ -0,0 +1,7175 @@
+	raster_info *		info;					/* pointer to rasterizer information */
+
+	INT16				ax, ay;					/* vertex A x,y (12.4) */
+	INT32				startr, startg, startb, starta; /* starting R,G,B,A (12.12) */
+	INT32				startz;					/* starting Z (20.12) */
+	INT64				startw;					/* starting W (16.32) */
+	INT32				drdx, dgdx, dbdx, dadx;	/* delta R,G,B,A per X */
+	INT32				dzdx;					/* delta Z per X */
+	INT64				dwdx;					/* delta W per X */
+	INT32				drdy, dgdy, dbdy, dady;	/* delta R,G,B,A per Y */
+	INT32				dzdy;					/* delta Z per Y */
+	INT64				dwdy;					/* delta W per Y */
+
+	INT64				starts0, startt0;		/* starting S,T (14.18) */
+	INT64				startw0;				/* starting W (2.30) */
+	INT64				ds0dx, dt0dx;			/* delta S,T per X */
+	INT64				dw0dx;					/* delta W per X */
+	INT64				ds0dy, dt0dy;			/* delta S,T per Y */
+	INT64				dw0dy;					/* delta W per Y */
+	INT32				lodbase0;				/* used during rasterization */
+
+	INT64				starts1, startt1;		/* starting S,T (14.18) */
+	INT64				startw1;				/* starting W (2.30) */
+	INT64				ds1dx, dt1dx;			/* delta S,T per X */
+	INT64				dw1dx;					/* delta W per X */
+	INT64				ds1dy, dt1dy;			/* delta S,T per Y */
+	INT64				dw1dy;					/* delta W per Y */
+	INT32				lodbase1;				/* used during rasterization */
+
+	UINT16				dither[16];				/* dither matrix, for fastfill */
+
+	Bitu				texcount;
+	Bitu				r_fbzColorPath;
+	Bitu				r_fbzMode;
+	Bitu				r_alphaMode;
+	Bitu				r_fogMode;
+	INT32				r_textureMode0;
+	INT32				r_textureMode1;
+};
+#endif
+
+/*************************************
+ *
+ *  Inline FIFO management
+ *
+ *************************************/
+
+inline INT32 fifo_space(fifo_state* f)
+{
+	INT32 items = 0;
+	if (items < 0)
+		items += f->size;
+	return f->size - 1 - items;
+}
+
+inline UINT8 count_leading_zeros(UINT32 value)
+{
+#ifdef _MSC_VER
+	DWORD idx = 0;
+	return (_BitScanReverse(&idx, value) ? (UINT8)(31 - idx) : (UINT8)32);
+#else
+	return (value ? (UINT8)__builtin_clz(value) : (UINT8)32);
+#endif
+	//INT32 result;
+	//
+	//#if defined _MSC_VER
+	//__asm
+	//{
+	//	bsr   eax,value
+	//	jnz   skip
+	//	mov   eax,63
+	//skip:
+	//	xor   eax,31
+	//	mov   result,eax
+	//}
+	//#else
+	//result = 32;
+	//while(value > 0) {
+	//	result--;
+	//	value >>= 1;
+	//}
+	//#endif
+	//return (UINT8)result;
+}
+
+/*************************************
+ *
+ *  Computes a fast 16.16 reciprocal
+ *  of a 16.32 value; used for
+ *  computing 1/w in the rasterizer.
+ *
+ *  Since it is trivial to also
+ *  compute log2(1/w) = -log2(w) at
+ *  the same time, we do that as well
+ *  to 16.8 precision for LOD
+ *  calculations.
+ *
+ *  On a Pentium M, this routine is
+ *  20% faster than a 64-bit integer
+ *  divide and also produces the log
+ *  for free.
+ *
+ *************************************/
+
+inline INT64 fast_reciplog(INT64 value, INT32* log2)
+{
+	UINT32 temp, rlog;
+	UINT32 interp;
+	UINT32 *table;
+	UINT64 recip;
+	bool neg = false;
+	int lz, exp = 0;
```

Comment:
**exp** is a function in cmath: `double exp(double x);`
Suggest renaming to avoid this shadowing.